pkg/daemon/build.go

```diff
@@ -36,3 +50,94 @@ func (srv *Daemon) buildHandler(engine api.Engine) func(w http.ResponseWriter, r
 		tgw.WriteResult(out)
 	}
 }
+
+func consumeBuildRequest(r *http.Request, dir string) (*api.BuildRequest, string, string, error) {
+	var (
+		req *api.BuildRequest
+		p   *multipart.Part
+		err error
+
+		plan string
+		sdk  string
+	)
+
+	if r.Body == nil {
+		return nil, "", "", fmt.Errorf("failed to parse request: nil body")
+	}
+
+	defer r.Body.Close()
+
+	// Validate the incoming multipart request.
+	ct, params, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
+	if err != nil {
+		return nil, "", "", fmt.Errorf("failed to parse request: %w", err)
+	}
+
+	if !strings.HasPrefix(ct, "multipart/") {
+		ct := r.Header.Get("Content-Type")
+		return nil, "", "", fmt.Errorf("not a multipart request; Content-Type: %s", ct)
+	}
+
+	rd := multipart.NewReader(r.Body, params["boundary"])
+
+	// 1. Read and decode the request payload.
+	if p, err = rd.NextPart(); err != nil {
+		return nil, "", "", fmt.Errorf("unexpected error when reading composition: %w", err)
+	}
+	if err = json.NewDecoder(p).Decode(&req); err != nil {
+		return nil, "", "", fmt.Errorf("failed to json decode request body: %w", err)
+	}
+
+	var (
+		planzip *os.File
+		sdkzip  *os.File
+	)
+
+	// 2a. Read test plan source archive.
```

Comment:
There is substantial storage redundancy here. For each run, a directory is created which contains the compressed and uncompressed copies of the plan being executed.